File: src/lib/t-books/dirty-discard.ts
import { exec, withTransaction } from "./db";

export function discardDirtyKey(kind: string, key: string): void {
  const k = key.trim();
  if (!k) return;
  withTransaction(() => {
    if (kind === "voucher") {
      const n = Number(k);
      if (!Number.isFinite(n)) return;
      exec("DELETE FROM voucher_payments WHERE voucher_number = ?", [n]);
      exec("DELETE FROM vouchers WHERE voucher_number = ?", [n]);
    } else if (kind === "purchase") {
      exec("DELETE FROM purchase_po WHERE po_number = ? COLLATE NOCASE", [k]);
    } else if (kind === "payment") {
      exec("DELETE FROM purchase_payments WHERE pay_number = ? COLLATE NOCASE", [k]);
    } else if (kind === "salary") {
      exec("DELETE FROM hr_payroll WHERE salary_number = ? COLLATE NOCASE", [k]);
    } else if (kind === "inventory") {
      const id = Number(k.replace(/^INV-/i, ""));
      if (Number.isFinite(id)) exec("DELETE FROM inventory WHERE id = ?", [id]);
    } else if (kind === "logistics") {
      const id = Number(k.replace(/^TRIP-/i, ""));
      if (Number.isFinite(id)) exec("DELETE FROM logistics WHERE id = ?", [id]);
    } else if (kind === "document") {
      const id = Number(k.replace(/^DOC-/i, ""));
      if (Number.isFinite(id)) exec("DELETE FROM documents WHERE id = ?", [id]);
    }
    try {
      exec("DELETE FROM pending_submit WHERE kind = ? AND key = ?", [kind, k]);
    } catch {
      /* table may be missing */
    }
  });
}

export function discardAllDirty(): void {
  withTransaction(() => {
    exec("DELETE FROM voucher_payments WHERE voucher_number IN (SELECT voucher_number FROM vouchers WHERE COALESCE(is_dirty, dirty, 0) = 1)");
    exec("DELETE FROM vouchers WHERE COALESCE(is_dirty, dirty, 0) = 1");
    try { exec("DELETE FROM purchase_po WHERE COALESCE(is_dirty,0) = 1"); } catch { /* */ }
    try { exec("DELETE FROM purchase_payments WHERE COALESCE(is_dirty,0) = 1"); } catch { /* */ }
    try { exec("DELETE FROM hr_payroll WHERE COALESCE(is_dirty,0) = 1"); } catch { /* */ }
    try { exec("DELETE FROM inventory WHERE COALESCE(is_dirty,0) = 1"); } catch { /* */ }
    try { exec("DELETE FROM logistics WHERE COALESCE(is_dirty,0) = 1"); } catch { /* */ }
    try { exec("DELETE FROM documents WHERE COALESCE(is_dirty,0) = 1"); } catch { /* */ }
    try { exec("DELETE FROM pending_submit"); } catch { /* */ }
  });
}

File: src-tauri/src/dirty_discard.rs
//! Drop local drafts that were never posted. Hive rows are not touched.
use rusqlite::params;

use crate::db::LocalBooks;
use crate::hive::{list_dirty_keys, DirtyKey, KIND_DOCUMENT, KIND_INVENTORY, KIND_LOGISTICS, KIND_PAYMENT, KIND_PURCHASE, KIND_SALARY, KIND_SALES_PO, KIND_VOUCHER};
use crate::Result;

fn clear_pending(books: &LocalBooks, kind: &str, key: &str) -> Result<()> {
    books.conn().execute(
        "DELETE FROM pending_submit WHERE kind = ?1 AND key = ?2",
        params![kind, key],
    )?;
    Ok(())
}

pub fn discard_dirty_key(books: &LocalBooks, kind: &str, key: &str) -> Result<()> {
    let kind = kind.trim();
    let key = key.trim();
    if kind.is_empty() || key.is_empty() {
        return Ok(());
    }
    match kind {
        k if k == KIND_VOUCHER => {
            let _ = books.conn().execute(
                "DELETE FROM voucher_payments WHERE CAST(voucher_number AS TEXT) = ?1",
                params![key],
            );
            books.conn().execute(
                "DELETE FROM vouchers WHERE CAST(voucher_number AS TEXT) = ?1 AND COALESCE(is_dirty, dirty, 0) = 1",
                params![key],
            )?;
        }
        k if k == KIND_PURCHASE => {
            books.conn().execute(
                "DELETE FROM purchase_po WHERE po_number = ?1 AND COALESCE(is_dirty,0) = 1",
                params![key],
            )?;
        }
        k if k == KIND_PAYMENT => {
            books.conn().execute(
                "DELETE FROM purchase_payments WHERE pay_number = ?1 AND COALESCE(is_dirty,0) = 1",
                params![key],
            )?;
        }
        k if k == KIND_SALES_PO => {
            let _ = books.conn().execute(
                "DELETE FROM sales_po WHERE po_number = ?1 AND COALESCE(is_dirty,0) = 1",
                params![key],
            );
        }
        k if k == KIND_SALARY => {
            books.conn().execute(
                "DELETE FROM hr_payroll WHERE salary_number = ?1 AND COALESCE(is_dirty,0) = 1",
                params![key],
            )?;
        }
        k if k == KIND_INVENTORY => {
            let id = key.trim_start_matches("INV-");
            books.conn().execute(
                "DELETE FROM inventory WHERE CAST(id AS TEXT) = ?1 AND COALESCE(is_dirty,0) = 1",
                params![id],
            )?;
        }
        k if k == KIND_LOGISTICS => {
            let id = key.trim_start_matches("TRIP-");
            books.conn().execute(
                "DELETE FROM logistics WHERE CAST(id AS TEXT) = ?1 AND COALESCE(is_dirty,0) = 1",
                params![id],
            )?;
        }
        k if k == KIND_DOCUMENT => {
            let id = key.trim_start_matches("DOC-");
            books.conn().execute(
                "DELETE FROM documents WHERE CAST(id AS TEXT) = ?1 AND COALESCE(is_dirty,0) = 1",
                params![id],
            )?;
        }
        _ => {}
    }
    clear_pending(books, kind, key)?;
    Ok(())
}

pub fn discard_all_dirty(books: &LocalBooks) -> Result<Vec<DirtyKey>> {
    let keys = list_dirty_keys(books)?;
    for k in &keys {
        discard_dirty_key(books, &k.kind, &k.key)?;
    }
    Ok(keys)
}
